### PR TITLE
HHH-15393 - Improve write-paths to use mapping model

### PR DESCRIPTION
**Pull request by sebersole**
HHH-15393 - Improve write-paths to use mapping model
HHH-15723 - Fix foreign-key modeling

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/annotations/Table.java [hunk 1]
```diff
@@ -87,10 +87,8 @@
 	 * If enabled, Hibernate will insert a row only if the columns of the secondary table
 	 * would not all be null, and will always use an outer join to read the columns. Thus,
 	 * by default, Hibernate avoids creating a row of null values.
-	 * <p>
-	 * <em>Only applies to secondary tables.</em>
 	 *
-	 * @deprecated use {@link SecondaryRow#optional()}
+	 * @apiNote Only relevant for secondary tables
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/annotations/Table.java:91]: Hey, what, I want `@Table` to go away, not resurrect it...

File: hibernate-core/src/main/java/org/hibernate/sql/exec/spi/JdbcOperationQuerySelect.java [hunk 2]
```diff
@@ -22,15 +22,15 @@
  *
  * @author Steve Ebersole
  */
-public class JdbcSelect extends AbstractJdbcOperation {
+public class JdbcOperationQuerySelect extends AbstractJdbcOperationQuery {
```
Comment (gavinking) [inline hibernate-core/src/main/java/org/hibernate/sql/exec/spi/JdbcOperationQuerySelect.java:25]: I'm not very keen on the naming of types in this hierarchy: "jdbc operation query select" reads like four words stuck together in a random order.  And "jdbc operation query delete" is even worse.

How about `SelectQueryOperation` or something? Or just `JdbcSelectOperation`?

**Comment by gavinking** [comment]
Wow. Epic commit!

File: hibernate-core/src/main/java/org/hibernate/annotations/Table.java (hunk 1 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/annotations/Table.java:91]: Which Jira did you use to remove it?

Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/annotations/Table.java:91]: https://github.com/hibernate/hibernate-orm/commit/aef9ab24257d58c85d1ee58add0f19f66b11867a#diff-52e80acb0c8fe3afd4721dec7bac45e2cf6141d20d76a91752e26da7c1143376

^ that's the last commit before mine.  You haven't deleted it there.  Am I misunderstanding you?

Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/annotations/Table.java:91]: Oh the deprecation.  Yeah it was a brutal rebase; I'm sure I missed stuff.  Did not help I had to sift through stuff 
 like brand new mutation operations (delete no version) as I was working on how mutations work ;)

Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/annotations/Table.java:91]: I fixed it locally.  Not going to push here though as I'm just waiting for CI to finish before pushing this upstream

File: hibernate-core/src/main/java/org/hibernate/sql/exec/spi/JdbcOperationQuerySelect.java (hunk 2 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/sql/exec/spi/JdbcOperationQuerySelect.java:25]: I agree the namings aren't great

File: hibernate-core/src/main/java/org/hibernate/annotations/Table.java (hunk 1 above)
Comment (gavinking) [reply hibernate-core/src/main/java/org/hibernate/annotations/Table.java:91]: yeah fine all good

File: hibernate-core/src/main/java/org/hibernate/sql/exec/spi/JdbcOperationQuerySelect.java (hunk 2 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/sql/exec/spi/JdbcOperationQuerySelect.java:25]: So the name has "Query" in it to indicate they come from Query objects.  Not sure a better way to indicate that - maybe moving the stuff under the `query` package...

File: hibernate-core/src/main/java/org/hibernate/engine/jdbc/spi/JdbcCoordinator.java [hunk 3]
```diff
@@ -35,13 +37,22 @@ public interface JdbcCoordinator extends Serializable, TransactionCoordinatorOwn
...
+
+	/**
+	 * Get the {@linkplain Batch batch} for the supplied key, creating one
+	 * if needed using the supplied {@linkplain PreparedStatementGroup statementGroupSupplier}.
 	 *
-	 * @return The batch
+	 * @implNote Any previous Batch is executed and released prior to returning
 	 */
-	Batch getBatch(BatchKey key);
+	Batch getBatch2(
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/engine/jdbc/spi/JdbcCoordinator.java:52]: Seems you missed a rename

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/DB2LegacySqlAstTranslator.java [hunk 4]
```diff
@@ -333,15 +332,6 @@ protected void visitUpdateStatementOnly(UpdateStatement statement) {
...
 	}
 
-	@Override
-	protected void visitInsertStatementOnly(InsertStatement statement) {
-		final boolean closeWrapper = renderReturningClause( statement );
-		super.visitInsertStatementOnly( statement );
-		if ( closeWrapper ) {
-			appendSql( ')' );
-		}
-	}
```
Comment (beikov) [inline hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/DB2LegacySqlAstTranslator.java:343]: I guess the removal of this was a victim of the renaming of `InsertStatement`? This should be aligned with `DB2SqlAstTranslator`.

File: hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/DB2LegacySqlAstTranslator.java [hunk 5]
```diff
@@ -429,11 +419,6 @@ protected String getFromDualForSelectOnly() {
 		return getFromDual();
 	}
 
-	@Override
-	protected void visitReturningColumns(MutationStatement mutationStatement) {
```
Comment (beikov) [inline hibernate-community-dialects/src/main/java/org/hibernate/community/dialect/DB2LegacySqlAstTranslator.java:433]: I guess the removal of this was a victim of the renaming of `InsertStatement`? This should be aligned with `DB2SqlAstTranslator`.

**Review by beikov — COMMENTED** [review]
Great work! I left a few comments, but overall, this all looks good to me!

**Pull request closed without merging**